mountpoint-s3-fs/src/manifest/builder.rs

```diff
@@ -1,62 +1,392 @@
+            // insert synthetic channel dir
+            let channel_root_id = self.next_id;
+            self.insert_buffer.push(DbEntry {
+                id: channel_root_id,
+                parent_id: FUSE_ROOT_ID,
+                channel_id,
+                parent_partial_key: None,
+                name: channel_manifest.directory_name.clone(),
+                etag: None,
+                size: None,
+            });
+            self.next_id += 1;
+
+            // insert keys from the manifest (and corresponding dirs)
+            for entry in channel_manifest.entries {
+                if let Err(ManifestError::FolderMarker(key)) = entry {
+                    tracing::warn!(
+                        "folder marker will be ignored: {}, channel directory may be empty: {}",
+                        key,
+                        channel_manifest.directory_name
+                    );
+                    continue;
+                }
+                let entry = entry?;
+
+                // insert the parent directories
+                let (parent_id, parent_partial_key) =
+                    self.ensure_dirs_inserted(&entry.partial_key, channel_id, channel_root_id)?;
+
+                // push new file entry to the insert_buffer
+                let db_entry = entry.into_db_entry(self.next_id, parent_id, channel_id, parent_partial_key);
+                self.insert_buffer.push(db_entry?);
+                self.next_id += 1;
+
+                // if insert_buffer is full, write to db
+                if self.insert_buffer.len() >= self.batch_size {
+                    self.flush_insert_buffer()?;
+                }
+            }
+
+            self.dir_ids.clear(); // dirs across channels do not overlap, forget them
+        }
+
+        self.flush_insert_buffer()?; // flush remaining entries to db
+
+        Ok(())
     }
 
-    db.create_index()?;
-    db.insert_directories(batch_size)?;
+    /// Ingests parent directory of `object_key` and its parents to db (if not done yet), returns ID and key of the immediate parent
+    ///
+    /// Note that insertion of directories is postponed till we have a full batch.
+    fn ensure_dirs_inserted(
+        &mut self,
+        object_key: &ValidKey,
+        channel_id: usize,
+        channel_root_id: u64,
+    ) -> Result<(u64, ValidKey), ManifestError> {
+        debug_assert_eq!(object_key.kind(), InodeKind::File);
 
-    Ok(())
-}
+        let components: Vec<_> = object_key.split('/').collect();
+        let mut parent_id = channel_root_id;
+        let mut parent_partial_key = ValidKey::root();
+
+        for component in components[..components.len() - 1].iter() {
+            let valid_name = ValidName::parse_str(component)
+                .map_err(|_| ManifestError::InvalidKey(ValidKeyError::InvalidKey(object_key.to_string())))?;
+            let partial_key = parent_partial_key.new_child(valid_name, InodeKind::Directory)?;
```

Comment:
If `object_key` is valid, aren't all its components also valid?

Reply:
that is true, `ValidName::parse_str(component)` is technically redundant here

but this is with the assumption that splitting is done right, I can simplify if we add the following method to `ValidKey`:

```rust
    /// Path components for this key.
    ///
    /// For directories, this does not include empty component after the terminal '/'.
    pub fn components(&self) -> Vec<ValidName> {
        self.key.trim_end_matches('/').split('/').map(ValidName).collect()
    }
```